compiler/src/vm/heap/mod.rs

```diff
@@ -0,0 +1,235 @@
+mod object;
+mod pointer;
+
+pub use self::{
+    object::{Builtin, Closure, Data, Int, Object, Struct, Symbol, Text},
+    pointer::Pointer,
+};
+use crate::builtin_functions::BuiltinFunction;
+use itertools::Itertools;
+use num_bigint::BigInt;
+use std::{cmp::Ordering, collections::HashMap};
+
+#[derive(Clone)]
+pub struct Heap {
+    objects: HashMap<Pointer, Object>,
+    next_address: Pointer,
+}
+
+impl std::fmt::Debug for Heap {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut objects = self.objects.clone().into_iter().collect_vec();
+        objects.sort_by_key(|(address, _)| address.raw());
+
+        writeln!(f, "{{")?;
+        for (address, object) in objects {
+            writeln!(
+                f,
+                "{address}: {} {}",
+                object.reference_count,
+                address.format(self),
+            )?;
+        }
+        write!(f, "}}")
+    }
+}
+
+impl Default for Heap {
+    fn default() -> Self {
+        Self {
+            objects: HashMap::new(),
+            next_address: Pointer::from_raw(1),
+        }
+    }
+}
+impl Heap {
+    pub fn get(&self, address: Pointer) -> &Object {
+        self.objects
+            .get(&address)
+            .unwrap_or_else(|| panic!("Couldn't get object {address}."))
+    }
+    pub fn get_mut(&mut self, address: Pointer) -> &mut Object {
+        self.objects
+            .get_mut(&address)
+            .unwrap_or_else(|| panic!("Couldn't get object {address}."))
+    }
+
+    pub fn dup(&mut self, address: Pointer) {
+        self.get_mut(address).reference_count += 1;
+        log::trace!(
+            "RefCount of {address} increased to {}. Value: {}",
+            self.get(address).reference_count,
+            address.format(self),
+        );
+    }
+    pub fn drop(&mut self, address: Pointer) {
+        let formatted_value = address.format(self);
+        let object = self.get_mut(address);
+        object.reference_count -= 1;
+        log::trace!(
+            "RefCount of {address} reduced to {}. Value: {formatted_value}",
+            object.reference_count,
+        );
+        if object.reference_count == 0 {
+            self.free(address);
+        }
+    }
+
+    pub fn create(&mut self, object: Data) -> Pointer {
+        let address = self.next_address;
+        self.objects.insert(
+            address,
+            Object {
+                reference_count: 1,
+                data: object,
+            },
+        );
+        log::trace!("Created object {} at {address}.", address.format(self));
```

Comment:
You could locally store the created `Object` and call it's `format(…)` directly to avoid the lookup

Reply:
Just to be clear: You propose formatting the object before moving it into the heap and then only outputting the string here? Or putting a clone of the object into the heap and then formatting the original in the trace?

Reply:
Either formatting it before it gets moved into the heap, or we just swap the order:

```rust
        let object = Object {
            reference_count: 1,
            data: object,
        };
        trace!("Creating object {} at {address}.", object.format(self));
        self.objects.insert(address, object);
        self.next_address = Pointer::from_raw(self.next_address.raw() + 1);
```